### PR TITLE
Websocket implementation

### PR DESCRIPTION
**Pull request by Witcher01**
following the esp_websocket_client found here[^1][^2], this commit
wraps the mentioned C library in Rust to expose a safe WebSocket
interface.
The exposed API copies the previously implemented MQTT API. On creating
a `EspWebSocketClientConfig` and passing it to a `EspWebSocketClient`,
the user can call `send` to send data. Manual sending of fragmented data,
Ping/Pong, or Connect/Disconnect are not supported due to the underlying
C library not supporting this, either.
Receiving is done in a new thread with the returned
`EspWebSocketConnection` and its blocking `next` function call, which
wakes on receiving one of the events defined in `WebSocketEventType`.
The `EspWebSocketClient` takes a `time::Duration` timeout since the
underlying C lib requires this but the defined interface in the
`embedded_svc` create does not provide it as a parameter.
On dropping the `EspWebSocketClient` the C functions
`esp_websocket_client_close` and `esp_websocket_client_destroy` are
being called, cleanly closing the connection to the WebSocket server.

[^1]: https://espressif.github.io/esp-protocols/esp_websocket_client/
[^2]: https://github.com/espressif/esp-protocols/

This requires esp-rs/esp-idf-sys/pull/88.

### CODE REVIEW

**Comment by ivmarkov** [comment]
Can you look at the build output. It fails not just with clippy, but also with real compilation errors?

**Comment by ivmarkov** [comment]
> Can you look at the build output. It fails not just with clippy, but also with real compilation errors?

Sorry, my mistake. I need to first merge the changes to `esp-idf-sys`.

**Comment by Witcher01** [comment]
@ivmarkov build failed due to `ws` being specified as an `experimental` feature in `embedded-svc`.
More errors follow after fixing this, which don't happen on the fork I develop on locally, but which can be reproduced when cloning cleanly. I'll investigate.

**Comment by ivmarkov** [comment]
Ease change the PR so that the experimental feature is temporarily enabled.

**Comment by ivmarkov** [comment]
You need to rebase your fork on top of latest esp-idf-svc master to resolve the conflicts.

**Comment by Witcher01** [comment]
@ivmarkov There are some warnings left in `src/eth.rs` which I suspect will prevent the `Build | Clippy` step in the CI from passing due to warnings being treated as errors.
Should I fix these, too?

**Comment by ivmarkov** [comment]
No it's OK like that. The eth related warnings seem to be for other architecture (xtens). Please keep in mind that my review will be delayed as I would be traveling tmr and over the weekend

File: src/ws/client.rs [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+    handle: esp_websocket_client_handle_t,
+    // used for the timeout in every call to a send method in the c lib as the
+    // `send` method in the `Sender` trait in embedded_svc::ws does not take a timeout itself
+    timeout: TickType_t,
+    #[allow(dead_code)]
+    callback: Box<dyn FnMut(i32, *mut esp_websocket_event_data_t)>,
+}
+
+impl EspWebSocketClient {
+    pub fn new(
```
Comment (ivmarkov) [inline src/ws/client.rs:406]: Rename to `new_with_connection` to match the latest mqtt API

File: src/ws/client.rs [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+            timeout,
+            Box::new(move |event_id, event_handle| {
+                EspWebSocketConnection::post(&client_connection, event_id, event_handle);
+            }),
+        )?;
+
+        Ok((client, connection))
+    }
+
+    fn new_with_raw_callback(
```
Comment (ivmarkov) [inline src/ws/client.rs:424]: Rename to just `new`.

File: src/ws/client.rs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+    ) -> Result<(), Self::Error> {
+        // NOTE: fragmented sending, as well as Closing or Continuing a connection, is not
+        // supported by the underlying C library and/or happen implicitly, e.g. when the
+        // `EspWebSocketClient` is dropped.
+        match frame_type {
+            FrameType::Binary(false) | FrameType::Text(false) => {
+                self.send_data(frame_type, frame_data)?
+            }
+            FrameType::Binary(true) | FrameType::Text(true) => {
+                panic!("Unsupported operation: Sending of fragmented data!")
```
Comment (ivmarkov) [inline src/ws/client.rs:544]: Perhaps remove the `!` from all messages?

**Review by ivmarkov — CHANGES_REQUESTED** [review]

File: src/ws/client.rs [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+
+        while message.is_some() {
+            message = self.0.processed.wait(message);
+        }
+    }
+
+    // NOTE: cannot implement the `Iterator` trait as it requires that all the items can be alive
+    // at the same time, which is not given here
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> Option<anyhow::Result<WebSocketEvent>> {
```
[the remark this answers is not in the document]
Comment (Witcher01) [reply src/ws/client.rs:376]: I'm not sure either, all I can say is that it does compile 😅 Should I change it to what you expected?

File: src/ws/client.rs [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+    pub uri: Option<&'a str>,
+    pub host: Option<&'a str>,
+    pub port: u16,
+    pub username: Option<&'a str>,
+    pub password: Option<&'a str>,
+    pub path: Option<&'a str>,
+    pub disable_auto_reconnect: bool,
+    // TODO: pub user_context:
+    pub task_prio: u8,
+    pub task_stack: u8,
```
[the remark this answers is not in the document]
Comment (Witcher01) [reply src/ws/client.rs:199]: I'm assuming you mean both `task_prio` and `task_stack` and I should cast it down to `u8` in the `try_from`?
Is that fine? The C lib requires a `u8`, but with a `usize` we're signalling that both variables can hold bigger values.

File: src/ws/client.rs [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+                        None
+                    })),
+                    _ => bail!(
+                        "WebSocket Data Event contains wrong OpCode: {}",
+                        event_data.op_code
+                    ),
+                }
+            }
+            esp_websocket_event_id_t_WEBSOCKET_EVENT_CLOSED => Ok(Self::Closed),
+            esp_websocket_event_id_t_WEBSOCKET_EVENT_MAX => {
```
[the remark this answers is not in the document]
Comment (Witcher01) [reply src/ws/client.rs:179]: I think you're right, I'm taking it out.

File: src/ws/client.rs [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+                bail!("Exceeded maximum packet size")
+            }
+            default => bail!("Unrecognized WebSocket event: {}", default),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct EspWebSocketClientConfig<'a> {
+    pub uri: Option<&'a str>,
```
[the remark this answers is not in the document]
Comment (Witcher01) [reply src/ws/client.rs:190]: I'm not sure what you mean with the last point. Should I just remove `username` and `password` from the config struct altogether?

**Pull request merged**